### PR TITLE
:sparkles: Add EVMC compatible state2

### PR DESCRIPTION
**Pull request by sheapwilliams**
Just for early review and feedback...

I've programmed a compatible EVMC state object based on state2.  I'm currently missing:

- mergeability tests (merge model is slightly different, so will require some adaptation)
- audit of existing *_state tests to make sure everything is accounted for
- in-order merging uses txn id currently, and that needs to evolve to the promise/future based approach.

It will require some adaptation in the other execution objects to work, but this is mostly around how a new revertible context is started/ended.

Lemme know your thoughts!

### CODE REVIEW

File: include/monad/state2/state.hpp [hunk 1]
```diff
@@ -1,52 +1,403 @@
...
+
+    void merge(State &new_state)
+    {
+        MONAD_DEBUG_ASSERT(can_merge(*this, new_state));
+
+        state::merge(*this, new_state);
+        accessed_ = std::move(new_state.accessed_);
+        accessed_storage_ = std::move(new_state.accessed_storage_);
+        total_selfdestructs_ = new_state.total_selfdestructs_;
+        gas_award_ = new_state.gas_award_;
```
Comment (sheapwilliams) [inline include/monad/state2/state.hpp]: Ignore this guy - totally wrong and currently in work...

File: include/monad/state2/state.hpp [hunk 2]
```diff
@@ -1,52 +1,403 @@
...
+        return account.has_value();
+    };
+
+    void create_account(address_t const &address)
+    {
+        MONAD_LOG_DEBUG(logger, "create_account: {}", address);
+
+        s.try_emplace(
+            address,
+            AccountState{.account = {std::nullopt, Account{}}, .storage = {}});
```
Comment (sheapwilliams) [inline include/monad/state2/state.hpp]: This guy isn't right - the account won't be in the BlockState, and subsequent storage accesses will fail.  But we don't want to fetch via `read_acount` - if there is a creation race and someone else wants to create the same account.

So we want to indicate a create explicitly, I think.  How do we either explicitly signal creation, or signal an abort if we detect somewhere we are going to collide?

File: include/monad/state2/state.hpp [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+        auto const previous_balance = account.value().balance;
+        MONAD_LOG_DEBUG(
+            logger,
+            "set_balance: {} = {}, ({}{})",
+            address,
+            intx::to_string(new_balance, 16),
+            new_balance >= previous_balance ? "+" : "-",
+            new_balance >= previous_balance
+                ? intx::to_string(new_balance - previous_balance, 16)
+                : intx::to_string(previous_balance - new_balance, 16));
```
Comment (zander-xyz) [inline include/monad/state2/state.hpp]: This is okay too, but we do have formatters for uint* values (see `monad/logging/formatter.hpp`)

File: include/monad/state2/state.hpp [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        }
+        return false;
+    }
+
+    [[nodiscard]] unsigned total_selfdestructs() const noexcept
+    {
+        return total_selfdestructs_;
+    }
+
+    void destruct_suicides() const noexcept {}
```
Comment (zander-xyz) [inline include/monad/state2/state.hpp:186]: I noticed that we also don't do anything in `state`, are we supposed to do something? Does this need an issue (and subsequent comment)?

File: include/monad/state2/state.hpp [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+
+    void merge(State &new_state)
+    {
+        s = std::move(new_state.s);
+        code_ = std::move(new_state.code_);
+        accessed_ = std::move(new_state.accessed_);
+        accessed_storage_ = std::move(new_state.accessed_storage_);
+        total_selfdestructs_ = new_state.total_selfdestructs_;
+        gas_award_ = new_state.gas_award_;
+        logs_ = std::move(new_state.logs_);
```
Comment (zander-xyz) [inline include/monad/state2/state.hpp]: Why are these moved as opposed to merged? Some detail of the ethereum spec?

File: src/monad/state2/test/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,12 @@
+add_unit_test(
+    # ENABLE_LOGGING
+    TARGET
+    monad-state2tests
+    SOURCES
+    #account_state.cpp
+    #code_state.cpp
```
Comment (zander-xyz) [inline src/monad/state2/test/CMakeLists.txt]: should these commented out files be removed completely?

File: include/monad/state2/state.hpp [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+
+MONAD_STATE_NAMESPACE_BEGIN
+
+// EVMC state object
+template <class Mutex, class TBlockCache>
+struct State
+{
+    BlockState<Mutex> &bs;
+    Db &db;
+    TBlockCache &block_cache_;
```
Comment (zander-xyz) [inline include/monad/state2/state.hpp:31]: When is the block_cache_ supposed to be written to? (i don't see it happening here)

File: include/monad/state2/state.hpp (hunk 5 above)
Comment (zander-xyz) [reply include/monad/state2/state.hpp]: okay I think I understand this part now...sorry

File: include/monad/state2/state.hpp (hunk 4 above)
Comment (sheapwilliams) [reply include/monad/state2/state.hpp:186]: Yea - this is an open issue.  We are supposed to only destruct suicides at the end of the txn...lemme find reference:

YP, Pg. 38 Virtual Machine Specification, Section f0s System Operations: 

> Halt execution and register account for later deletion.

Register for later deletion is the key, as the account still exists for purposes of gathering reward, etc.  But when deleted, anything left over is blackholed.  Turning into an issue.

File: include/monad/state2/state.hpp (hunk 7 above)
Comment (sheapwilliams) [reply include/monad/state2/state.hpp:31]: It gets written to internally, if you request a valid block header hash but haven't read it yet (cause you started up not from the genesis block), then it will record it so that it is cached.

In normal operation, the cache will be populated as blocks are written, and the state won't ever have a non-const interface to this block.

We can make this const by making this populate all the block header hashes on startup, and then the only non-const usage will be in sync with the block processor...
